### PR TITLE
Support data subsampling

### PR DESCRIPTION
**Pull request by fehiepsi**
Addresses #556. This will also be helpful for #724.

@fritzo The implementation here is a bit simpler than Pyro. But I am not sure if I miss any point. Could you help me review this PR?

### CODE REVIEW

File: numpyro/primitives.py [hunk 1]
```diff
@@ -224,38 +239,44 @@ def __init__(self, name, size, subsample_size=None, dim=None):
...
             raise ValueError('dim arg must be negative.')
-        self.dim = dim
-        self._validate_and_set_dim()
+        self.dim, self._indices = self._subsample(self.name, self.size, self.subsample_size, dim)
         super(plate, self).__init__()
 
-    def _validate_and_set_dim(self):
+    # XXX: different from Pyro, this method returns dim and indices
+    @staticmethod
+    def _subsample(name, size, subsample_size, dim):
```
Comment (fehiepsi) [inline numpyro/primitives.py:247]: I made this a staticmethod to be able to reuse it in `contrib.funsor`. The api is also more similar to Pyro.

File: numpyro/contrib/funsor/enum_messenger.py [hunk 2]
```diff
@@ -442,10 +442,11 @@ def __init__(self, name, size, subsample_size=None, dim=None):
...
             raise ValueError('dim arg must be negative.')
         self.dim = dim
+        _, indices = OrigPlateMessenger._subsample(self.name, self.size, self.subsample_size, dim)
         self._indices = funsor.Tensor(
-            funsor.ops.new_arange(funsor.tensor.get_default_prototype(), self.size),
-            OrderedDict([(self.name, funsor.bint(self.size))]),
-            self.size
+            indices,
+            OrderedDict([(self.name, funsor.bint(self.subsample_size))]),
+            self.subsample_size
```
Comment (fehiepsi) [inline numpyro/contrib/funsor/enum_messenger.py:449]: @eb8680 I made those changes to support subsampling. Could you take a look to see if this is a reasonable change?

File: numpyro/handlers.py [hunk 3]
```diff
@@ -596,7 +599,7 @@ def __init__(self, fn=None, data=None, substitute_fn=None):
         super(substitute, self).__init__(fn)
 
     def process_message(self, msg):
-        if (msg['type'] not in ('sample', 'param')) or msg.get('_control_flow_done', False):
+        if (msg['type'] not in ('sample', 'param', 'plate')) or msg.get('_control_flow_done', False):
```
Comment (fehiepsi) [inline numpyro/handlers.py:602]: I added `plate` here so we can substitute if needed. This is useful for #724 I guess. The behavior should be similar to the way we use `plate(..., subsample=)` in Pyro.

File: numpyro/handlers.py [hunk 4]
```diff
@@ -547,7 +547,10 @@ def __init__(self, fn=None, rng_seed=None):
 
     def process_message(self, msg):
         if (msg['type'] == 'sample' and not msg['is_observed'] and
-                msg['kwargs']['rng_key'] is None) or msg['type'] == 'control_flow':
+                msg['kwargs']['rng_key'] is None) or msg['type'] in ['plate', 'control_flow']:
+            # no need to create a new key when value is available
+            if msg['value'] is not None:
+                return
```
Comment (fehiepsi) [inline numpyro/handlers.py:553]: This saves some unnecessary splitting random keys.

File: numpyro/contrib/funsor/enum_messenger.py (hunk 2 above)
Comment (eb8680) [reply numpyro/contrib/funsor/enum_messenger.py:449]: Looks right to me. The corresponding code in `pyro.contrib.funsor` is here: https://github.com/pyro-ppl/pyro/blob/dev/pyro/contrib/funsor/handlers/plate_messenger.py#L63

**Comment by eb8680** [comment]
Looks mostly correct to me, but you might want to add tests that exercise replay and nested/reused plates, like the model in [this Pyro test](https://github.com/pyro-ppl/pyro/blob/dev/tests/infer/test_valid_models.py#L681).

> The implementation here is a bit simpler than Pyro. But I am not sure if I miss any point.

Part of the extra complexity in Pyro comes from supporting sequential plates, which I guess isn't a goal here, and part of it comes from treating subsample sites as `sample` sites, which was just a poor early design choice in Pyro. Also, are you planning to add a [`numpyro.subsample` primitive](http://docs.pyro.ai/en/stable/primitives.html#pyro.primitives.subsample)?

**Comment by fehiepsi** [comment]
> add tests that exercise replay and nested/reused plates

Thanks @eb8680 ! I didn't even think about this. Thanks for pointing it out. :)

> planning to add a numpyro.subsample primitive?

I didn't have that plan. Let me see if it is easy to address.

**Comment by eb8680** [comment]
> Let me see if it is easy to address.

The implementation in Pyro is here in `SubsampleMessenger._postprocess_message`: https://github.com/pyro-ppl/pyro/blob/dev/pyro/poutine/subsample_messenger.py#L134

Here's a simplified version that does not include parameter subsampling:
```py
    def postprocess_message(self, msg):  # new method for plate
        if msg["type"] == "subsample" and self.dim is not None:
            event_dim = msg["kwargs"].get("event_dim")
            if event_dim is not None:
                assert event_dim >= 0
                dim = self.dim - event_dim
                shape = msg["value"].shape
                if len(shape) >= -dim and shape[dim] != 1:
                    if is_validation_enabled() and shape[dim] != self.size:
                        statement = "pyro.subsample(..., event_dim={})".format(event_dim)
                        raise ValueError(
                            "Inside pyro.plate({}, {}, dim={}) invalid shape of {}: {}"
                            .format(self.name, self.size, self.dim, statement, shape))
                    if self.subsample_size < self.size:
                        value = msg["value"]
                        new_value = value.index_select(dim, self._indices)
                        msg["value"] = new_value
```

**Comment by fehiepsi** [comment]
Thanks for the snippet, Eli! My intention is just making a quick PR for this but it is worth to extend this to support subsample and param primitives. Also Pyro allows subsample_size is None in model and not None in guide. I remember @fritzo said that he will use NumPyro if we have this feature implemented. That's a good motivation. :))

File: numpyro/contrib/funsor/enum_messenger.py [hunk 5]
```diff
@@ -438,14 +439,14 @@ class plate(GlobalNamedMessenger):
     def __init__(self, name, size, subsample_size=None, dim=None):
         self.name = name
         self.size = size
-        self.subsample_size = size if subsample_size is None else subsample_size
         if dim is not None and dim >= 0:
             raise ValueError('dim arg must be negative.')
-        self.dim = dim
+        self.dim, indices = OrigPlateMessenger._subsample(self.name, self.size, subsample_size, dim)
```
Comment (fehiepsi) [inline numpyro/contrib/funsor/enum_messenger.py:444]: I changed the behavior here: set `self.dim` to be the output of `_subsample`, instead of default value in the constructor.

**Comment by fehiepsi** [comment]
@eb8680 I have added `subsample` primitive and support subsample for it and `param`. Also made a small change to allow `subsample_size=None` in `model` but not None in `guide`.

**Review by fritzo — COMMENTED** [review]
Note sure if this first subsample PR is the right place for thorough tests of SVI, but I found Pyro's [test_subsample_gradient()](https://github.com/pyro-ppl/pyro/blob/90bbb3e4314a7b7d2cba6f36a440aee521681dda/tests/infer/test_gradient.py#L40) to be invaluable in detecting subsampling issues with very little compute cost. In particular it helped us correctly scale the different parts of the `ScoreParts` tuple.

Does this PR aim to support subsampling in SVI? If so do you already have good tests for `poutine.scale()`?

File: numpyro/primitives.py [hunk 6]
```diff
@@ -221,41 +241,50 @@ class plate(Messenger):
...
-                break
-            dim -= 1
-        if self.dim is None:
-            self.dim = dim
+        if dim is None:
+            new_dim = -1
+            while True:
+                if new_dim not in occupied_dims:
+                    break
+                new_dim -= 1
```
[the remark this answers is not in the document]
Comment (fehiepsi) [reply numpyro/primitives.py]: Thanks!

**Comment by fehiepsi** [comment]
@fritzo Thanks for reviewing! I just add `subsample_gradient` test. We have some tests for `handlers.scale` in `test_mask` and `test_scale`, but we haven't supported `ScoreParts` stuff in NumPyro yet.

> Does this PR aim to support subsampling in SVI?

My main motivation is to support https://github.com/pyro-ppl/numpyro/issues/724, where subsample is required (see [this comment](https://github.com/pyro-ppl/numpyro/issues/724#issuecomment-688491882)). But SVI should work I believe.

File: setup.py [hunk 7]
```diff
@@ -51,8 +51,8 @@
             'ipython',
             'isort',
             'flax',
-            'dm-haiku @ https://github.com/deepmind/dm-haiku/archive/v0.0.2.zip',
-            'tfp-nightly',  # TODO: change this to stable release or a specific nightly release
+            'dm-haiku',
+            'tfp-nightly==0.12.0.dev20200911',  # TODO: change this to stable release or a specific nightly release
```
Comment (fehiepsi) [inline setup.py:55]: I merged #739 to here because without it, all PRs will fail.

File: numpyro/handlers.py (hunk 4 above)
Comment (fritzo) [reply numpyro/handlers.py:553]: So do I understand correctly that in numpyro, in contrast to pyro, `plate` does not consume randomness? Note in Pyro it is `plate` that creates a `_Subsample` site that consumes randomness.

File: test/test_handlers.py [hunk 8]
```diff
@@ -235,6 +258,102 @@ def test_plate(model):
...
+    num_particles = 50000
+    optimizer = optim.Adam(0.1)
+    elbo = ELBO(num_particles=num_particles)
+    svi = SVI(model, guide, optimizer, loss=elbo)
+    svi_state = svi.init(random.PRNGKey(0), None)
+    params = svi.optim.get_params(svi_state.optim_state)
+    normalizer = 2 if subsample else 1
+    if subsample_size == 1:
+        subsample = jnp.array([0])
+        _, grads1 = value_and_grad(lambda x: svi.loss.loss(
```
Comment (fritzo) [inline test/test_handlers.py:340]: could we also check that the values agree, since it's not much extra work?

**Review by fritzo — COMMENTED** [review]
Looks great, thanks for implementing this!

File: numpyro/handlers.py (hunk 4 above)
Comment (fehiepsi) [reply numpyro/handlers.py:553]: @fritzo IIUC your question, then in NumPyro, `_subsample_fn` plays a similar role. If `value != None`, we will use the random key here to generate subsample indices.

Comment (fritzo) [reply numpyro/handlers.py:553]: thanks for explaining!

File: numpyro/primitives.py [hunk 9]
```diff
@@ -202,13 +211,24 @@ def module(name, nn, input_shape=None):
     return functools.partial(nn_apply, nn_params)
 
 
+def _subsample_fn(size, subsample_size, rng_key=None):
+    assert rng_key is not None, "Missing random key to generate subsample indices."
+    return random.permutation(rng_key, size)[:subsample_size]
```
[the remark this answers is not in the document]
Comment (fehiepsi) [reply numpyro/primitives.py:216]: I think you did some [bechmarks](https://github.com/pyro-ppl/pyro/pull/2383) for this. In JAX, both are [equivalent](https://github.com/google/jax/blob/master/jax/random.py#L592) so using `random.choice` is also good to me.

[the remark this answers is not in the document]
Comment (fritzo) [reply numpyro/primitives.py:216]: Ha ha thanks for finding the evidence. No need for change then.

File: test/test_handlers.py (hunk 8 above)
Comment (fehiepsi) [reply test/test_handlers.py]: Hi @fritzo, I just added a test for value agreement. But I am not sure if that is what you mentioned.

**Review by fritzo — APPROVED** [review]

**Comment by fritzo** [comment]
LGTM, @eb8680 did you have any other comments?

**Review by eb8680 — DISMISSED** [review]
Looks good, thanks for adding the extra tests!

**Comment by fehiepsi** [comment]
Thanks for reviewing, @eb8680 and @fritzo! I just observed something wrong in `Predictive` for models with multiple `subsample` statements. I'll try to fix it and add the failing test in this PR.

File: test/test_handlers.py [hunk 10]
```diff
@@ -235,6 +258,110 @@ def test_plate(model):
...
+    data = jnp.arange(100.)
+    subsample_size = 7
+
+    with handlers.trace() as guide_trace, handlers.seed(rng_seed=0):
+        with numpyro.plate("a", len(data), subsample_size=subsample_size):
+            pass
+
+    with handlers.seed(rng_seed=1), handlers.replay(guide_trace=guide_trace):
+        with numpyro.plate("a", len(data)):
+            subsample_data = numpyro.subsample(data, event_dim=0)
```
Comment (fehiepsi) [inline test/test_handlers.py:303]: This test exercises `subsample` works under `replay` handler.

File: numpyro/handlers.py [hunk 11]
```diff
@@ -192,7 +192,7 @@ def __init__(self, fn=None, guide_trace=None):
         super(replay, self).__init__(fn)
 
     def process_message(self, msg):
-        if msg['name'] in self.guide_trace and msg['type'] in ('sample', 'plate'):
+        if msg['type'] in ('sample', 'plate') and msg['name'] in self.guide_trace:
```
Comment (fehiepsi) [inline numpyro/handlers.py:195]: I switch the order of two sides of `and` because some primitives such as `subsample` do not have `name` field.

**Review by fritzo — APPROVED** [review]
